core/src/channel_group/config.rs

```diff
@@ -0,0 +1,90 @@
+use crate::{channel::ChannelInitOptions, AudioStreamParams};
+
+/// Defines the multithreading options for each task that supports it.
+#[derive(Clone)]
+pub enum ThreadCount {
+    /// No multithreading. Run everything on the same thread.
+    None,
+
+    /// Run with multithreading, with an automatically determined thread count.
+    /// Please read
+    /// [this](https://docs.rs/rayon-core/1.5.0/rayon_core/struct.ThreadPoolBuilder.html#method.num_threads)
+    /// for more information about the thread count selection.
+    Auto,
+
+    /// Run with multithreading, with the specified thread count.
+    Manual(usize),
+}
+
+/// Options regarding which parts of the ChannelGroup should be multithreaded.
+///
+/// Responsibilities of a channel: processing input events for the channel,
+/// dispatching per-key rendering of audio, applying filters to the final channel's audio
+///
+/// Responsibilities of a key: Rendering per-voice audio for all the voices stored in a
+/// key for a channel. This is generally the most compute intensive part of the synth.
+///
+/// Best practices:
+/// - As there are often 16 channels in MIDI, per-key multithreading can balance out the
+///     load more evenly between CPU cores.
+/// - However, per-key multithreading adds some overhead, so if the synth is invoked to
+///     render very small sample counts each time (e.g. sub 1 millisecond), not using per-key
+///     multithreading becomes more efficient.
+#[derive(Clone)]
+pub struct ParallelismOptions {
+    /// Render the MIDI channels parallel in a threadpool with the specified
+    /// thread count.
+    pub channel: ThreadCount,
+
+    /// Render the individisual keys of each channel parallel in a threadpool
+    /// with the specified thread count.
+    pub key: ThreadCount,
+}
+
+impl ParallelismOptions {
+    pub const AUTO_PER_KEY: Self = ParallelismOptions {
+        channel: ThreadCount::None,
```

Comment:
Should be auto here still, because if you're parallelizing voice processing per key you'd probably also want to parallelize event processing too (which must happen per channel)

```suggestion
        channel: ThreadCount::Auto,
```

Reply:
True, will change this and also use this const for Default. Is it ok to merge after this change?

Reply:
yeah